### PR TITLE
Implement rooms for LBP1 clients

### PR DESCRIPTION
**Pull request by Beyley**

### CODE REVIEW

File: Refresh.GameServer/Endpoints/Game/Handshake/AuthenticationEndpoints.cs [hunk 1]
```diff
@@ -261,18 +262,21 @@ private static void SendPlatformNotAllowedNotification(GameDatabaseContext datab
...
     [GameEndpoint("goodbye", HttpMethods.Post, ContentType.Xml)]
     [MinimumRole(GameUserRole.Restricted)]
-    public Response RevokeThisToken(RequestContext context, GameDatabaseContext database, GameUser user)
+    public Response RevokeThisToken(RequestContext context, GameDatabaseContext database, GameUser user, Token token, MatchService matchService)
     {
-        string? token = context.Cookies["MM_AUTH"];
+        // On LBP1, our concept of a "room" is very basic, and we can be sure that they are the only user in said "room",
+        // so we just remove it outright if the user is logging out
+        // NOTE: We don't need any extra logic to handle more complex rooms here,
+        //       since LBP2 and LBP3 do not actually send requests to this endpoint when closed.
```
Comment (jvyden) [inline Refresh.GameServer/Endpoints/Game/Handshake/AuthenticationEndpoints.cs:270]: I think they can, but the games usually segfault before the game can actually send this request. Best to not leave this as a special case. Only condition should be if the user is the host of that room.

File: Refresh.GameServer/Endpoints/Game/MatchingEndpoints.cs [hunk 2]
```diff
@@ -37,4 +37,33 @@ public Response Match(RequestContext context, GameDatabaseContext database, Game
...
+            room.LevelType = RoomSlotType.Pod;
+            room.LevelId = id;
+        }
+        else
+        {
+            room.LevelType = slotType == "user" ? RoomSlotType.Online : RoomSlotType.Story;
+            room.LevelId = id;
+        }
+        
+        room.LastContact = DateTimeOffset.Now;
```
Comment (jvyden) [inline Refresh.GameServer/Endpoints/Game/MatchingEndpoints.cs:63]: Should probably make all requests update this field on LBP1. Or maybe just `/notification`; the only ticking request LBP1 makes.

File: Refresh.GameServer/Endpoints/Game/MatchingEndpoints.cs [hunk 3]
```diff
@@ -37,4 +37,33 @@ public Response Match(RequestContext context, GameDatabaseContext database, Game
...
     }
+    
+    // Sent by LBP1 to notify the server it has entered a level.
+    // While it doesn't send us any detailed room information, we can at least create a pseudo-room on the server with just that player.
+    // Due to it only sending this when the level is *entered*, it means its much more likely for the room to be auto cleared due to inactivity,
+    // since the "bump" is much less often. This will at the very least make API tools be able to see LBP1 player activity and player counts
+    [GameEndpoint("enterLevel/{slotType}/{id}", HttpMethods.Post)]
+    public Response EnterLevel(RequestContext context, Token token, MatchService matchService, string slotType, int id)
+    {
+        GameRoom room = matchService.GetOrCreateRoomByPlayer(token.User, token.TokenPlatform, token.TokenGame, NatType.Moderate);
```
Comment (jvyden) [inline Refresh.GameServer/Endpoints/Game/MatchingEndpoints.cs:48]: Should assume worst-case scenario
```suggestion
        GameRoom room = matchService.GetOrCreateRoomByPlayer(token.User, token.TokenPlatform, token.TokenGame, NatType.Strict);
```

File: Refresh.GameServer/Endpoints/Game/Handshake/AuthenticationEndpoints.cs [hunk 4]
```diff
@@ -261,18 +262,21 @@ private static void SendPlatformNotAllowedNotification(GameDatabaseContext datab
     /// </summary>
     [GameEndpoint("goodbye", HttpMethods.Post, ContentType.Xml)]
     [MinimumRole(GameUserRole.Restricted)]
-    public Response RevokeThisToken(RequestContext context, GameDatabaseContext database, GameUser user)
+    public Response RevokeThisToken(RequestContext context, GameDatabaseContext database, GameUser user, Token token, MatchService matchService)
     {
-        string? token = context.Cookies["MM_AUTH"];
+        // On LBP1, our concept of a "room" is very basic, and we can be sure that they are the only user in said "room",
+        // so we just remove it outright if the user is logging out
```
Comment (jvyden) [inline Refresh.GameServer/Endpoints/Game/Handshake/AuthenticationEndpoints.cs:268]: Comments probably don't need to exist.

**Review by jvyden — CHANGES_REQUESTED** [review]

File: Refresh.GameServer/Endpoints/Game/MatchingEndpoints.cs (hunk 2 above)
Comment (Beyley) [reply Refresh.GameServer/Endpoints/Game/MatchingEndpoints.cs:63]: `/notification` is a good idea, i'll do that

File: Refresh.GameServer/Endpoints/Game/Handshake/AuthenticationEndpoints.cs (hunk 1 above)
Comment (Beyley) [reply Refresh.GameServer/Endpoints/Game/Handshake/AuthenticationEndpoints.cs:270]: Even when the games cleanly exited (tried on both PS3 and RPCS3, i could get the games to safely close) i couldn't get either to send this request, but I'll extend the logic to handle it anyway

Comment (jvyden) [reply Refresh.GameServer/Endpoints/Game/Handshake/AuthenticationEndpoints.cs:270]: I'd prefer it be correct even if the games are stupid.

File: Refresh.GameServer/Endpoints/Game/Handshake/AuthenticationEndpoints.cs [hunk 5]
```diff
@@ -262,23 +262,17 @@ private static void SendPlatformNotAllowedNotification(GameDatabaseContext datab
     /// </summary>
     [GameEndpoint("goodbye", HttpMethods.Post, ContentType.Xml)]
     [MinimumRole(GameUserRole.Restricted)]
-    public Response RevokeThisToken(RequestContext context, GameDatabaseContext database, GameUser user, Token token, MatchService matchService)
+    public Response RevokeThisToken(RequestContext context, GameDatabaseContext database, Token token, MatchService matchService)
```
Comment (jvyden) [inline Refresh.GameServer/Endpoints/Game/Handshake/AuthenticationEndpoints.cs:265]: Keep the reference to the user; removing it is bound to cause issues when we revoke the token in this method.

**Review by jvyden — CHANGES_REQUESTED** [review]

File: Refresh.GameServer/Endpoints/Game/Handshake/AuthenticationEndpoints.cs (hunk 5 above)
Comment (Beyley) [reply Refresh.GameServer/Endpoints/Game/Handshake/AuthenticationEndpoints.cs]: Fixed in #79f24aab24d93bf4a03729004fd2cbf27b22203d

**Review by jvyden — APPROVED** [review]

**Pull request merged**